### PR TITLE
Run PR validation check only when needed & use docker cache during build

### PR DESCRIPTION
**Pull request by IgorGov**

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -19,14 +19,16 @@ jobs:
     name: Unit Tests
     runs-on: ubuntu-latest
     steps:
+      - name: Check out code into the Go module directory
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 2
```
Comment (gadotroee) [inline .github/workflows/test.yml:25]: why do we need to fetch-depth 2?

**Review by gadotroee — COMMENTED** [review]
one small question that I want to understand but it looks good

**Comment by gadotroee** [comment]
also, I see it caused the codecov to fail and I guess it is because we are not uploading the full report (for example all exceptions tests didn't run here).. 
do we have any solution for it?

File: .github/workflows/test.yml (hunk 1 above)
Comment (IgorGov) [reply .github/workflows/test.yml:25]: for git diff to work

**Comment by IgorGov** [comment]
> also, I see it caused the codecov to fail and I guess it is because we are not uploading the full report (for example all exceptions tests didn't run here)..
> do we have any solution for it?
no.

File: .github/workflows/test.yml (hunk 1 above)
Comment (gadotroee) [reply .github/workflows/test.yml:25]: 👍

**Review by RamiBerm — APPROVED** [review]

**Review by gadotroee — APPROVED** [review]

**Pull request merged**